docs/CustomCodeTemplates.md

```diff
@@ -0,0 +1,167 @@
+# Custom Code Templates
+
+CleverTap Unity SDK supports a custom presentation of in-app messages. This allows for utilizing the in-app notifications functionality with custom configuration and presentation logic. Two types of templates can be defined through the SDK: Templates and Functions. Templates can contain action arguments while Functions cannot and Functions can be used as actions while Templates cannot. Functions can be either "visual" or "non-visual". "Visual" functions can contain UI logic and will be part of the [In-App queue](#in-App-queue), while "non-visual" functions will be triggered directly when invoked and should not contain UI logic.
+
+## Creating templates and functions
+Templates consist of a name, type, and arguments (and isVisual for Functions). Type and arguments (and isVisual for Functions) are required and names must be unique across the application. The template definitions are validated for correctness and will throw `CustomTemplateException` when an invalid template is created on launching the application. Those exceptions should generally not be caught and template definitions must be valid to be triggered correctly.
+
+The templates are defined in а JSON format with the following scheme:
+```json
+{
+  "TemplateName": {
+    "type": "template",
+    "arguments": {
+      "Argument1": {
+        "type": "string|number|boolean|file|action|object",
+        "value": "val"
+        },
+      "Argument2": {
+        "type": "object",
+        "value": {
+          "Nested1": {
+            "type": "string|number|boolean|object",
+            "value": "val"
+          },
+          "Nested2": {
+            "type": "string|number|boolean|object",
+            "value": "val"
+          }
+        }
+      }
+    }
+  },
+  "functionName": {
+    "type": "function",
+    "isVisual": true|false,
+    "arguments": {
+      "a": {
+      "type": "string|number|boolean|file|object",
+      "value": "val"
+      }
+    }
+  }
+}
+```
+The JSON definitions should be placed in one or more files in `Assets/CleverTap/CustomTemplates`.
```

Comment:
I believe information on where the files are copied to and how they're handled on iOS/Android will be helpful. We can add this later on in a separate PR once the export script is finalized.